v2alpha1/pkg/api/api.proto

```diff
@@ -0,0 +1,472 @@
+  Container container = 2;
+}
+
+message CreateContainerResponse {
+  // Requested adjustments to container being created.
+  ContainerAdjustment adjust = 1;
+  // Requested updates to other existing containers.
+  repeated ContainerUpdate update = 2;
+  // Requested eviction of existing containers.
+  repeated ContainerEviction evict = 3;
+}
+
+message StartContainerEvent {
+  // Pod of container being started.
+  PodSandbox pod = 1;
+  // Container being started.
+  Container container = 2;
+}
+
+message UpdateContainerRequest {
+  // Pod of container being updated.
+  PodSandbox pod = 1;
+  // Container being updated.
+  Container container = 2;
+  // Resources to update.
+  LinuxResources linux_resources = 3;
+}
+
+message UpdateContainerResponse {
+  // Requested updates to containers.
+  repeated ContainerUpdate update = 1;
+  // Requested eviction of containers.
+  repeated ContainerEviction evict = 2;
+}
+
+message StopContainerRequest {
+  // Pod of container being stopped.
+  PodSandbox pod = 1;
+  // Container being stopped.
+  Container container = 2;
+}
+
+message StopContainerResponse {
+  // Requested updates to containers.
+  repeated ContainerUpdate update = 1;
+}
+
+message RemoveContainerEvent {
+  // Pod of container being removed.
+  PodSandbox pod = 1;
+  // Container being removed.
+  Container container = 2;
+}
+
+message PostCreateContainerEvent {
+  // Pod of container created.
+  PodSandbox pod = 1;
+  // Container created.
+  Container container = 2;
+}
+
+message PostStartContainerEvent {
+  // Pod of container started.
+  PodSandbox pod = 1;
+  // Container started.
+  Container container = 2;
+}
+
+message PostUpdateContainerEvent {
+  // Pod of container updated.
+  PodSandbox pod = 1;
+  // Container updated.
+  Container container = 2;
+}
+
+// Empty response for those *Requests that are semantically events.
+message Empty {}
+
+// Events that plugins can subscribe to in ConfigureResponse.
+enum Event {
+  RUN_POD_SANDBOX = 0;
+  STOP_POD_SANDBOX = 1;
+  REMOVE_POD_SANDBOX = 2;
+  CREATE_CONTAINER = 3;
+  POST_CREATE_CONTAINER = 4;
+  START_CONTAINER = 5;
+  POST_START_CONTAINER = 6;
+  UPDATE_CONTAINER = 7;
+  POST_UPDATE_CONTAINER = 8;
+  STOP_CONTAINER = 9;
+  REMOVE_CONTAINER = 10;
+  ALL = 11;
+}
+
+// Pod metadata that is considered relevant for a plugin.
+message PodSandbox {
+  string id = 1;
+  string name = 2;
+  string uid = 3;
+  string namespace = 4;
+  map<string, string> labels = 5;
+  map<string, string> annotations = 6;
+  string cgroup_parent = 7;
+  string runtime_handler = 8;
+}
+
+// Container metadata that is considered relevant for a plugin.
+message Container {
+  string id = 1;
+  string pod_sandbox_id = 2;
+  string name = 3;
+  ContainerState state = 4;
+  map<string, string> labels = 5;
+  map<string, string> annotations = 6;
+  repeated string args = 7;
+  repeated string env = 8;
+  repeated Mount mounts = 9;
+  Hooks hooks = 10;
+  LinuxContainer linux = 11;
+}
```

Comment:
Can we just pass OCI `config.json` as string here?
Requests to update the config can be represented as JSON patches

https://en.wikipedia.org/wiki/JSON_Patch

Reply:
Well, it would definitely be possible technically. The primary reason why we decided against it so far is the scope of changes we want to allow vs. the full OCI Spec scope. 

Since we want to restrict changes via NRI to a relatively small subset of the OCI Spec, it felt more natural (and easier) to bolt most of the restrictions into the protocol itself. IOW, to have the protocol itself already prevent 'syntactically' one from expressing most of the disallowed changes as opposed to have a protocol that 'syntactically' allows arbitrary changes which then need to be 'semantically' restricted with a bunch of extra code.

A minor/secondary reason is related to the 'API-ish' aspects of the protocol. Using JSON patches on the OCI Spec seems awkward as an API to me, but this is just my personal subjective opinion. This point also might be a bit moot, since we prefer to see the plugin stub as the primary API towards plugin writers instead of the low-level ttRPC protocol itself. If the low-level protocol was designed with a full OCI-Spec input/JSON patch output approach, this fact could be hidden from the user by/in the stub as well.